lib/crawler/keyword/keywords.ex

```diff
@@ -8,28 +8,71 @@ defmodule Crawler.Keyword.Keywords do
   alias Crawler.Keyword.Queries.KeywordQuery
   alias Crawler.Keyword.Schemas.Keyword
   alias Crawler.Repo
+  alias CrawlerWorker.Keyword.CrawlerWorker
 
   def list_keywords(user_id), do: Repo.all(KeywordQuery.user_keywords_query(user_id))
 
+  def get_keyword_by_id(id), do: Repo.get(Keyword, id)
+
   def create_keyword(attrs \\ %{}) do
     %Keyword{}
     |> Keyword.changeset(attrs)
     |> Repo.insert()
   end
 
   def create_keyword_list(keyword_list, user_id) do
-    keyword_list
-    |> build_keyword_params_list(user_id)
-    |> insert_all_keywords()
+    create_keyword_fn = fn k ->
+      case create_keyword(k) do
+        {:ok, %{id: keyword_id}} -> keyword_id
+        # throw  error if a keyword is invalid
+        {:error, changeset} -> throw({:error, changeset})
+      end
+    end
+
+    Repo.transaction(fn ->
+      try do
+        keyword_ids =
+          keyword_list
+          |> build_keyword_params_list(user_id)
+          |> Enum.map(create_keyword_fn)
+
+        enqueue_crawling_job(keyword_ids)
+
+        keyword_ids
+      catch
+        # TODO: handle to show error
+        {:error, _changeset} -> Repo.rollback(:keyword_length_exceeded)
+      end
+    end)
   end
 
-  defp insert_all_keywords(keyword_params_list), do: Repo.insert_all(Keyword, keyword_params_list)
+  def mark_as_in_progress(keyword) do
+    keyword
+    |> Keyword.in_progress_changeset()
+    |> Repo.update()
+  end
 
-  defp build_keyword_params_list(keyword_list, user_id) do
-    now = NaiveDateTime.truncate(NaiveDateTime.utc_now(), :second)
+  def mark_as_completed(keyword, attrs) do
+    keyword
+    |> Keyword.completed_changeset(attrs)
+    |> Repo.update()
+  end
+
+  def mark_as_failed(keyword) do
+    keyword
+    |> Keyword.failed_changeset()
+    |> Repo.update()
+  end
 
+  defp build_keyword_params_list(keyword_list, user_id) do
     Enum.map(keyword_list, fn keyword ->
-      %{name: keyword, status: :new, user_id: user_id, inserted_at: now, updated_at: now}
+      %{name: keyword, user_id: user_id}
     end)
   end
+
+  defp enqueue_crawling_job(keyword_ids) do
+    keyword_ids
+    |> Enum.map(fn id -> CrawlerWorker.new(%{"keyword_id" => id}) end)
+    |> Oban.insert_all()
```

Comment:
Does it mean that we will start to process all keywords at the same time? (well, 10 keywords start at the same time, and the next keywords will be processed once one of the previous was finished).

There could be issues with google limits.

Also, there is a line in IC requirements addresses this issue:

>You need to find a way to work around the limitations of mass-searching keywords as Google prevents it. Be creative.

Reply:
Oh I'm new to this tbh. I haven't faced google limits yet. Would we prefer to add the job one by one?
As @andyduong1920's [suggestion](https://github.com/vnntsu/elixir-ic/pull/39#discussion_r1063292503), I try to use `insert_all` and I think it's a better performance for inserting jobs into DB.
I'm not sure about the google search, @andyduong1920 @Nihisil What should we go with?

Reply:
1/ Insert here is just `insert the data into Oban database`
2/ And then, how does Oban will process it later -> that is another topic
3/ About the google limit, we could have some approach like
- Limit the queue, like only 1 request per second, Oban has support via `queue, limit and the dispatch_cooldown`

I used the `dispatch_cooldown` on my IC FYI https://github.com/andyduong1920/ex_marketer/blob/development/config/config.exs#L44

Reply:
@andyduong1920 Nice! I've just stolen yours. Thanks for the detail.
Fixed in [cfe4051](https://github.com/vnntsu/elixir-ic/pull/39/commits/cfe4051a683bac5435d6caa2b47eb50dc095e728)

Reply:
I suggest to upload 1000 keywords file (since IC requirement is support such files) and see how current limit will work :slightly_smiling_face: